### ISSUE TITLE
On the Result ID Conversion and Memory Consumption of Zero Sample Cell Annotation

### ISSUE THREAD

**Issue by haolangchen** [bowang-lab/scGPT#216]
Hello, I would like to use zero samples model  (best_model. pt about 33M human) to directly predict cell types without fine-tuning the model

However, I found that the final result only outputs the id (1, 2, 3) for the celltype

![image](https://github.com/bowang-lab/scGPT/assets/128439741/40bb1bd0-6a01-4ad6-9365-1a82cbaa6227)

How can I convert the output ID into a cell type to type through a dictionary

Can you provide an the idtotype dictionary for the zero sample model (best_model. pt)

In addition, I run best.model. pt on 7000 cells and found that the memory consumption was very high. I want to know where the problem lies. Is this memory consumption normal。
Using GPUs still requires significant resources

![39a0210f9d143b9088f7d819d2c34030](https://github.com/bowang-lab/scGPT/assets/128439741/0f9d0caa-8923-46bb-ac09-10e811c50560)

**Comment by subercui** [comment]
Hi, if you want to use the model in the zero-shot way, you may see this tutorial https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Reference_Mapping.ipynb. In summary, the notebook search for nearest neighbors based on cell embeddings, and propagate the label of the neighbors in a reference atlas to your new cells.

For this part:
> outputs the id (1, 2, 3) for the celltype
I guess you probably used some other tutorials. So sometimes there is a new classifier appended to the model, and that is the output of the classifier. You can ignore these "cls_output" if you want to use the model in zero-shot manner, instead, please look at the tutorial mentioned above.

Regarding resource consumption, can you show the GPU consumption by nvidia-smi? It is possible you may be still running the code on CPUs

**Comment by haolangchen** [comment]
Hello, thank you very much for your prompt reply。
The cell annotation tutorial I am using is https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Annotation.ipynb

What I want to ask is, is the tutorial I am using only applicable to Fine tuning on Pre trained Model for Cell type Annotation
And cannot directly use the model in zero-shot manner for prediction

My key code is as follows, without making any model Fine-tuning 

  ```
###设定模型基本参数
model = TransformerModel(
    ntokens,
    embsize,
    nhead,
    d_hid,
    nlayers,
    nlayers_cls=3,
    n_cls=num_types if CLS else 1,
    vocab=vocab,
    dropout=dropout,
    pad_token=pad_token,
    pad_value=pad_value,
    do_mvc=MVC,
    do_dab=DAB,
    use_batch_labels=INPUT_BATCH_LABELS,
    num_batch_labels=num_batch_types,
    domain_spec_batchnorm=config["DSBN"],
    input_emb_style=input_emb_style,
    n_input_bins=n_input_bins,
    cell_emb_style=cell_emb_style,
    mvc_decoder_style=mvc_decoder_style,
    ecs_threshold=ecs_threshold,
    explicit_zero_prob=explicit_zero_prob,
    use_fast_transformer=fast_transformer,
    fast_transformer_backend=fast_transformer_backend,
    pre_norm=config["pre_norm"],
)



if config["load_model"] is not None:
    try:
        model.load_state_dict(torch.load(model_file))
        print(f"Loading all model params from {model_file}")
    except:
        # only load params that are in the model and match the size
        model_dict = model.state_dict()
        pretrained_dict = torch.load(model_file)
        pretrained_dict = {
            k: v
            for k, v in pretrained_dict.items()
            if k in model_dict and v.shape == model_dict[k].shape
        }
        for k, v in pretrained_dict.items():
            print(f"Loading params {k} with shape {v.shape}")
        model_dict.update(pretrained_dict)
        model.load_state_dict(model_dict)



#计算了模型中所有可训练参数的数量，但排除了不需要梯度的参数。
pre_freeze_param_count = sum(dict((p.data_ptr(), p.numel()) for p in model.parameters() if p.requires_grad).values())

predictions = []
  model.eval()
  with torch.no_grad():
      with torch.cuda.amp.autocast(enabled=True):
          output_dict = model(
                          tokenized_train["genes"],
                          tokenized_train["values"],
                          src_key_padding_mask=src_key_padding_mask,
                          batch_labels=batch_ids if INPUT_BATCH_LABELS or config["DSBN"] else None,
                          CLS=False,
                          CCE=False,
                          MVC=False,
                          ECS=False,
                          do_sample=do_sample_in_train,
                          #generative_training=False
                      )
      
       output_values = output_dict["mlm_output"]
      preds = output_values.argmax(1).cpu().numpy()
      predictions.append(preds)
  
  
  print(predictions)

```

If you suggest that I use another tutorial, then I have another question.

In your article, it was mentioned that the prediction accuracy of fine-tuning models is higher than that of zero-shot models。

So, are the zero-shot models you use for annotation also using (https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Reference_Mapping.ipynb) tutorials？

If you are using the methods in the (https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Annotation.ipynb) tutorial to evaluate the accuracy of fine-tuning models and zero-shot models, why can't I directly use the  zero-shot model for predictions and convert ID to celltype, or why can't I attach an ID conversion dictionary to the download list for us to use。

I am looking forward to your reply again

**Comment by haolangchen** [comment]
> Hi, if you want to use the model in the zero-shot way, you may see this tutorial https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Reference_Mapping.ipynb. In summary, the notebook search for nearest neighbors based on cell embeddings, and propagate the label of the neighbors in a reference atlas to your new cells.
> 
> For this part:
> 
> > outputs the id (1, 2, 3) for the celltype
> > I guess you probably used some other tutorials. So sometimes there is a new classifier appended to the model, and that is the output of the classifier. You can ignore these "cls_output" if you want to use the model in zero-shot manner, instead, please look at the tutorial mentioned above.
> 
> Regarding resource consumption, can you show the GPU consumption by nvidia-smi? It is possible you may be still running the code on CPUs


Hello, thank you very much for your prompt reply。
The cell annotation tutorial I am using is https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Annotation.ipynb

What I want to ask is, is the tutorial I am using only applicable to Fine tuning on Pre trained Model for Cell type Annotation
And cannot directly use the model in zero-shot manner for prediction

My key code is as follows, without making any model Fine-tuning

```
###设定模型基本参数
model = TransformerModel(
  ntokens,
  embsize,
  nhead,
  d_hid,
  nlayers,
  nlayers_cls=3,
  n_cls=num_types if CLS else 1,
  vocab=vocab,
  dropout=dropout,
  pad_token=pad_token,
  pad_value=pad_value,
  do_mvc=MVC,
  do_dab=DAB,
  use_batch_labels=INPUT_BATCH_LABELS,
  num_batch_labels=num_batch_types,
  domain_spec_batchnorm=config["DSBN"],
  input_emb_style=input_emb_style,
  n_input_bins=n_input_bins,
  cell_emb_style=cell_emb_style,
  mvc_decoder_style=mvc_decoder_style,
  ecs_threshold=ecs_threshold,
  explicit_zero_prob=explicit_zero_prob,
  use_fast_transformer=fast_transformer,
  fast_transformer_backend=fast_transformer_backend,
  pre_norm=config["pre_norm"],
)



if config["load_model"] is not None:
  try:
      model.load_state_dict(torch.load(model_file))
      print(f"Loading all model params from {model_file}")
  except:
      # only load params that are in the model and match the size
      model_dict = model.state_dict()
      pretrained_dict = torch.load(model_file)
      pretrained_dict = {
          k: v
          for k, v in pretrained_dict.items()
          if k in model_dict and v.shape == model_dict[k].shape
      }
      for k, v in pretrained_dict.items():
          print(f"Loading params {k} with shape {v.shape}")
      model_dict.update(pretrained_dict)
      model.load_state_dict(model_dict)



#计算了模型中所有可训练参数的数量，但排除了不需要梯度的参数。
pre_freeze_param_count = sum(dict((p.data_ptr(), p.numel()) for p in model.parameters() if p.requires_grad).values())

predictions = []
model.eval()
with torch.no_grad():
    with torch.cuda.amp.autocast(enabled=True):
        output_dict = model(
                        tokenized_train["genes"],
                        tokenized_train["values"],
                        src_key_padding_mask=src_key_padding_mask,
                        batch_labels=batch_ids if INPUT_BATCH_LABELS or config["DSBN"] else None,
                        CLS=False,
                        CCE=False,
                        MVC=False,
                        ECS=False,
                        do_sample=do_sample_in_train,
                        #generative_training=False
                    )
    
     output_values = output_dict["mlm_output"]
    preds = output_values.argmax(1).cpu().numpy()
    predictions.append(preds)


print(predictions)
```

If you suggest that I use another tutorial, then I have another question.

In your article, it was mentioned that the prediction accuracy of fine-tuning models is higher than that of zero-shot models。

So, are the zero-shot models you use for annotation also using (https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Reference_Mapping.ipynb) tutorials？

If you are using the methods in the (https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Annotation.ipynb) tutorial to evaluate the accuracy of fine-tuning models and zero-shot models, why can't I directly use the zero-shot model for predictions and convert ID to celltype, or why can't I attach an ID conversion dictionary to the download list for us to use。

I am looking forward to your reply again

**Comment by haolangchen** [comment]
> Hi, if you want to use the model in the zero-shot way, you may see this tutorial https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Reference_Mapping.ipynb. In summary, the notebook search for nearest neighbors based on cell embeddings, and propagate the label of the neighbors in a reference atlas to your new cells.
> 
> For this part:
> 
> > outputs the id (1, 2, 3) for the celltype
> > I guess you probably used some other tutorials. So sometimes there is a new classifier appended to the model, and that is the output of the classifier. You can ignore these "cls_output" if you want to use the model in zero-shot manner, instead, please look at the tutorial mentioned above.
> 
> Regarding resource consumption, can you show the GPU consumption by nvidia-smi? It is possible you may be still running the code on CPUs

Hello, thank you very much for your prompt reply。
The cell annotation tutorial I am using is https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Annotation.ipynb

What I want to ask is, is the tutorial I am using only applicable to Fine tuning on Pre trained Model for Cell type Annotation
And cannot directly use the model in zero-shot manner for prediction

My key code is as follows, without making any model Fine-tuning

```
###设定模型基本参数
model = TransformerModel(
  ntokens,
  embsize,
  nhead,
  d_hid,
  nlayers,
  nlayers_cls=3,
  n_cls=num_types if CLS else 1,
  vocab=vocab,
  dropout=dropout,
  pad_token=pad_token,
  pad_value=pad_value,
  do_mvc=MVC,
  do_dab=DAB,
  use_batch_labels=INPUT_BATCH_LABELS,
  num_batch_labels=num_batch_types,
  domain_spec_batchnorm=config["DSBN"],
  input_emb_style=input_emb_style,
  n_input_bins=n_input_bins,
  cell_emb_style=cell_emb_style,
  mvc_decoder_style=mvc_decoder_style,
  ecs_threshold=ecs_threshold,
  explicit_zero_prob=explicit_zero_prob,
  use_fast_transformer=fast_transformer,
  fast_transformer_backend=fast_transformer_backend,
  pre_norm=config["pre_norm"],
)



if config["load_model"] is not None:
  try:
      model.load_state_dict(torch.load(model_file))
      print(f"Loading all model params from {model_file}")
  except:
      # only load params that are in the model and match the size
      model_dict = model.state_dict()
      pretrained_dict = torch.load(model_file)
      pretrained_dict = {
          k: v
          for k, v in pretrained_dict.items()
          if k in model_dict and v.shape == model_dict[k].shape
      }
      for k, v in pretrained_dict.items():
          print(f"Loading params {k} with shape {v.shape}")
      model_dict.update(pretrained_dict)
      model.load_state_dict(model_dict)



#计算了模型中所有可训练参数的数量，但排除了不需要梯度的参数。
pre_freeze_param_count = sum(dict((p.data_ptr(), p.numel()) for p in model.parameters() if p.requires_grad).values())

predictions = []
model.eval()
with torch.no_grad():
    with torch.cuda.amp.autocast(enabled=True):
        output_dict = model(
                        tokenized_train["genes"],
                        tokenized_train["values"],
                        src_key_padding_mask=src_key_padding_mask,
                        batch_labels=batch_ids if INPUT_BATCH_LABELS or config["DSBN"] else None,
                        CLS=False,
                        CCE=False,
                        MVC=False,
                        ECS=False,
                        do_sample=do_sample_in_train,
                        #generative_training=False
                    )
    
     output_values = output_dict["mlm_output"]
    preds = output_values.argmax(1).cpu().numpy()
    predictions.append(preds)


print(predictions)
```
If you suggest that I use another tutorial, then I have another question.

In your article, it was mentioned that the prediction accuracy of fine-tuning models is higher than that of zero-shot models。

So, are the zero-shot models you use for annotation also using (https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Reference_Mapping.ipynb) tutorials？

If you are using the methods in the (https://github.com/bowang-lab/scGPT/blob/main/tutorials/Tutorial_Annotation.ipynb) tutorial to evaluate the accuracy of fine-tuning models and zero-shot models, why can't I directly use the zero-shot model for predictions and convert ID to celltype, or why can't I attach an ID conversion dictionary to the download list for us to use。

I am looking forward to your reply again
